singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -272,450 +129,636 @@ PORTABLE_INLINE_FUNCTION bool solve_Ax_b(Real *a, Real *b) {
-      ComponentEnergies[i] *= (TotalSIE * TotalMass / Esum);
+class CacheAccessor {
+ public:
+  PORTABLE_INLINE_FUNCTION
+  CacheAccessor() = default;
+  PORTABLE_INLINE_FUNCTION
+  explicit CacheAccessor(Real *scr) : cache_(scr) {}
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real * operator[](const int m) const {
+    return cache_ + m*MAX_NUM_LAMBDAS;
   }
+ private:
+  Real *cache_;
+};
 
-  // This is the main loop
-  constexpr const int pte_max_iter = nmat * pte_max_iter_per_mat;
-  for (iter = 0; iter < pte_max_iter; ++iter) {
-    // Call the EOSs
-    for (int mat = 0; mat < nmat; ++mat) {
-      Densities[mat] = ComponentMasses[mat] / ComponentVolumes[mat];
-      eoss[ofst(mat)].PTofRE(Densities[mat], ComponentEnergies[mat], lambdas[mat],
-                             Pressures[mat], Temperatures[mat], dpdr[mat], dpde[mat],
-                             dtdr[mat], dtde[mat]);
+template <typename EOSIndexer, typename RealIndexer>
+class PTESolverBase {
+ public:
+  PTESolverBase() = delete;
+  PORTABLE_INLINE_FUNCTION int Nmat() const { return nmat; }
+  PORTABLE_INLINE_FUNCTION int &Niter() { return niter; }
+  PORTABLE_INLINE_FUNCTION
+  virtual void Fixup() const {
+    Real vsum = 0;
+    for (int m = 0; m < nmat; ++m) vsum += vfrac[m];
+    for (int m = 0; m < nmat; ++m) vfrac[m] *= vfrac_total/vsum;
+  }
+  PORTABLE_INLINE_FUNCTION
+  void Finalize() {
+    for (int m = 0; m < nmat; m++) {
+      temp[m] *= Tnorm;
+      u[m] *= u_total;
+      press[m] *= u_total;
     }
-    // Now, zero the matrix and vector
-    for (int i = 0; i < 2 * nmat * 2 * nmat; ++i)
-      A[i] = 0.0;
-    for (int i = 0; i < 2 * nmat; ++i)
-      b[i] = 0.0;
-    // print_matrix(2*nmat,2*nmat,A);
-    // print_vector(2*nmat,b);
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  bool Solve() const {
+    for (int m = 0; m < neq; m++) dx[m] = residual[m];
+    return solve_Ax_b_wscr(neq, jacobian, dx, sol_scratch);
+  }
+ protected:
+  PORTABLE_INLINE_FUNCTION
+  PTESolverBase(int nmats, int neqs, EOSIndexer &eos_, const Real vfrac_tot,
+                const Real sie_tot, RealIndexer &rho_, RealIndexer &vfrac_,
+                RealIndexer &sie_, RealIndexer &temp_, RealIndexer &press_,
+                Real *&scratch)
+    : nmat(nmats), neq(neqs), niter(0),
+      eos(eos_),
+      vfrac_total(vfrac_tot),
+      sie_total(sie_tot),
+      rho(rho_),
+      vfrac(vfrac_),
+      sie(sie_),
+      temp(temp_),
+      press(press_) {
+    jacobian = AssignIncrement(scratch, neq*neq);
+    dx = AssignIncrement(scratch, neq);
+    sol_scratch = AssignIncrement(scratch, 2*neq);
+    residual = AssignIncrement(scratch, neq);
+    u = AssignIncrement(scratch, nmat);
+    rhobar = AssignIncrement(scratch, nmat);
+    Cache = CacheAccessor(AssignIncrement(scratch, nmat * MAX_NUM_LAMBDAS));
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  void InitBase() {
+    Real Tguess = 0.0;
+    for (int m = 0; m < nmat; ++m) Tguess = std::max(Tguess, temp[m]);
+    Tnorm = Tguess;
```

Comment:
As a follow-up, this appears to have units of temperature whereas it appears to be used as if it is unitless.

Reply:
Yeah, the input temperatures need to not be nonsense.  zero is fine for some of the materials, like if they didn't exist in a cell before a particular pte call, but they shouldn't just be random garbage.  we could make a different choice.  for example, we could just always guess 300K or something to start, but my preference is just to force callers to have at least initialized everything with something sane.  The benefit of this is that hosts can provide excellent initial guesses with little effort (like the cell's previous temperature) which can have a large impact on PTE solver performance.  same can be said of density/volume fractions.  In our block prototype code, I'm seeing much less than one iteration of the PTE solver on average for simple problems because the initial guesses are excellent.

Reply:
> As a follow-up, this appears to have units of temperature whereas it appears to be used as if it is unitless.

I'm not sure what you're referring to here.

Reply:
> > As a follow-up, this appears to have units of temperature whereas it appears to be used as if it is unitless.
> 
> I'm not sure what you're referring to here.

There are several places in the code like this
```c++
temp[m] *= Tnorm;
```
or
```c++
eos[m].RhoPmin(Tnorm * Tideal)
```
or
```c++
sie[m] = eos[m].InternalEnergyFromDensityTemperature(rho[m], Tnorm * Tideal, Cache[m]);
```
or
```c++
Real e_pert = eos[m].InternalEnergyFromDensityTemperature(rho_pert, Tnorm * Tequil, Cache[m]);
```
These expressions in my mind imply that either `Tnorm` or what it multiplies should be unitless. Looking at the code more carefully, it looks like `Tideal` may be unitless?
```c++
  PORTABLE_INLINE_FUNCTION
  void GetIdealPTE(Real &Pideal, Real &Tideal) const {
    Real rhoBsum = 0.0;
    Real Asum = 0.0;
    for (int m = 0; m < nmat; m++) {
      Asum += vfrac[m] * press[m] / temp[m];
      rhoBsum += rho[m] * vfrac[m] * sie[m] / temp[m];
    }
    Asum *= u_total/Tnorm;
    rhoBsum /= Tnorm;
    Tideal = u_total / rhoBsum / Tnorm;
    Pideal = Tideal * Asum / vfrac_total / u_total;
  }
```
  
But `Tequil` certainly looks to have units, right? I see it initialized here with units for example:
```c++
PORTABLE_INLINE_FUNCTION
Real Init() {
  InitBase();
  Residual();
  TryIdealPTE(this);
  Tequil = temp[0];
  return ResidualNorm();
}
```

Reply:
Okay so I think I understand my confusion a bit more. `temp[:]` is normalized basically everywhere, hence multiplying it by `Tnorm` in the `Finalize()` member function. However, I would assume that somewhere there should be a
```c++
for (int m = 0; m < nmat; m++) {temp[m] /= Tnorm;}
```
somewhere in the initialization. Am I missing it?

Reply:
Can we add some (debug only) checks that enforce that the input temperatures are not nonsense? I.e., check for negative, NaN or infinite inputs?